integration_tests/run.sh

```diff
@@ -28,61 +28,194 @@
+         "g2g_tests=FAILED" "g2g_documentation=FAILED" "g2g_package=geo2grid-${suffix}"
+    elif [[ "${GIT_TAG_NAME:0:3}" = "p2g" ]] || [[ "$commit_message" =~ (^|.[[:space:]])"["p2g(-skip-tests)?"]"$ ]]; then
+        prefixes=polar
+        save_vars "p2g_tests=FAILED" "p2g_documentation=FAILED" "p2g_package=polar2grid-${suffix}"\
+         "g2g_tests=SKIPPED" "g2g_documentation=SKIPPED" "g2g_package="
+    else
+        prefixes="geo polar"
+        save_vars "p2g_tests=FAILED" "p2g_documentation=FAILED" "p2g_package=polar2grid-${suffix}"\
+         "g2g_tests=FAILED" "g2g_documentation=FAILED" "g2g_package=geo2grid-${suffix}"
+    fi
+}
+
+setup_conda()
+{
+    # Activate conda for bash.
+    /data/users/davidh/miniconda3/bin/conda init bash
+    # Restart the shell to enable conda.
+    source ~/.bashrc
+
+    conda env update -n jenkins_p2g_swbundle -f "$WORKSPACE"/build_environment.yml
+    # Documentation environment also has behave, while the build environment does not.
+    conda env update -n jenkins_p2g_docs -f "$WORKSPACE"/build_environment.yml -f "${WORKSPACE}/jenkins_environment.yml"
+    conda activate jenkins_p2g_docs
+    pip install -U --no-deps "$WORKSPACE"
+}
+
+format_test_details()
+{
+    test_details="$1"
+    json_file="${WORKSPACE}/integration_tests/json_file.txt"
+    # Gets the line before json data starts.
+    i=`grep -n "^\[$" "$test_details" | grep -oE "[0-9]+"`
+    # Gets the line after json data ends.
+    j=`grep -n "^\]$" "$test_details" | grep -oE "[0-9]+"`
+    # Remove lines that are not json data.
+    sed "1,${i}d;${j},\$d" "$test_details" > "$json_file"
+    set +x
+    python << EOF > "$test_details"
+import json
+with open("json_file.txt") as json_file:
+    data = json.load(json_file)
+    print()
+    for test in data['elements']:
+        name = test['name'].split('@')[1]
+        duration = 0
+        for step in test['steps']:
+            duration += step['result']['duration'] if step.get('result') else 0
+        end = '\n'
+        if test == data['elements'][-1]:
+            end = ''
+        print("\t\t{0}: {1} in {2} seconds".format(name, test['status'], round(duration)), end=end)
+EOF
+    set -x
+    rm "$json_file"
+}
+
+run_tests()
+{
+    set +e
+    (
+        set -e
+        export POLAR2GRID_HOME="$swbundle_name"
+        prefix=$1
+        test_details="${WORKSPACE}/integration_tests/${prefix:0:1}2g_test_details.txt"
+        json_file="${WORKSPACE}/integration_tests/json_file.txt"
+        cd "${WORKSPACE}/integration_tests"
+        # Prints output to stdout and to an output file.
+        behave --no-logcapture --no-color --no-capture -D datapath=/data/test_data -i "${prefix}2grid.feature"\
+         --format pretty --format json.pretty 2>&1 | tee "$test_details"
+
+        format_test_details "$test_details"
+
+        # Replace FAILED with SUCCESSFUL.
+        save_vars "${prefix:0:1}2g_tests=SUCCESSFUL"
+    )
+}
+
+create_documentation()
+{
+    set +e
+    (
```

Comment:
Could you explain what you're trying to do with this `set +e` followed by the parentheses?

Reply:
Essentially a "try" clause. I want it to break out of that segment if an error occurs without breaking out of the entire program. This allows finish_time to be made, AND allow documentation to be tested even if the tests failed.

Reply:
Would you rather have the documentation fail if tests fail? I think there should be a try clause around the main part of the program so finish_time is saved regardless of any error. I can remove the other try clauses of you think the documentation should fail if tests fail.

Reply:
Wow, really cool! I think the way you have it is good then. We'll change it in the future if it becomes a problem. We don't want documentation to be generated and **uploaded** if tests fail, but generating it alone should be fine.